lib/mix/test/mix/tasks/do_test.exs

```diff
@@ -24,4 +24,127 @@ defmodule Mix.Tasks.DoTest do
+
+        assert_received {:mix_shell, :info,
+                         [
+                           "warning: running \"help\" at root level because it is not a recursive task (ignoring apps options)"
+                         ]}
+      end)
+    end)
+  end
+
+  test "runs given tasks ignoring apps argument when project is not an umbrella" do
+    Mix.Tasks.Do.run(["--app", "bar", "--app", "foo", "cmd", "echo", "hello"])
+
+    assert_received {:mix_shell, :info,
+                     [
+                       "warning: running \"cmd\" at root level because this is not an umbrella project"
+                     ]}
+  end
+
+  test "runs given aliases for each app specified by app flag" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      aliases = [
+        e: ["cmd echo hello"],
+        p: fn val -> Mix.shell().info(inspect(val)) end
+      ]
+
+      Mix.Project.in_project(:umbrella, ".", [aliases: aliases], fn _ ->
+        Mix.Tasks.Do.run(["compile"])
+        Mix.Tasks.Do.run(["--app", "bar", "--app", "foo", "e,", "p", "Foo"])
+
+        nl = os_newline()
+        assert_received {:mix_shell, :info, ["==> bar"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+        assert_received {:mix_shell, :info, ["[\"Foo\"]"]}
+
+        assert_received {:mix_shell, :info, ["==> foo"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+        assert_received {:mix_shell, :info, ["[\"Foo\"]"]}
+      end)
+    end)
+  end
+
+  test "runs non-recursive aliases at project root level" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      aliases = [
+        h: "help"
+      ]
+
+      Mix.Project.in_project(:umbrella, ".", [aliases: aliases], fn _ ->
+        Mix.Tasks.Do.run(["--app", "bar", "--app", "foo", "h"])
+
+        assert_received {:mix_shell, :info,
+                         [
+                           "warning: running \"help\" at root level because it is not a recursive task (ignoring apps options)"
+                         ]}
+      end)
+    end)
+  end
+
+  test "runs given aliases ignoring apps argument when project is not an umbrella" do
+    in_fixture("umbrella_dep/deps/umbrella", fn ->
+      aliases = [
+        e: ["cmd echo hello"],
+        p: fn val -> Mix.shell().info(inspect(val)) end
+      ]
+
+      Mix.Project.in_project(:foo, "apps/foo", [aliases: aliases], fn _ ->
+        Mix.Tasks.Do.run(["--app", "bar", "--app", "foo", "e,", "p", "Foo"])
+
+        assert_received {:mix_shell, :info,
+                         [
+                           "warning: running \"cmd\" at root level because this is not an umbrella project"
+                         ]}
+
+        assert_received {:mix_shell, :info,
+                         [
+                           "warning: running \"p\" at root level because this is not an umbrella project"
+                         ]}
+      end)
+    end)
+  end
```

Comment:
I need to confirm we have all cases (code branches) covered with these test cases.